compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirWhenConditionChecker.kt

```diff
@@ -21,5 +29,42 @@ object FirWhenConditionChecker : FirWhenExpressionChecker() {
                 checkCondition(condition, context, reporter)
             }
         }
+        if (expression.subject != null) {
+            checkDuplicatedLabels(expression, context, reporter)
+        }
+    }
+
+    private fun checkDuplicatedLabels(expression: FirWhenExpression, context: CheckerContext, reporter: DiagnosticReporter) {
+        // The second part of each pair indicates whether the `is` check is positive or negated.
+        val checkedTypes = HashSet<Pair<ConeKotlinType, FirOperation>>()
+        val checkedConstants = HashSet<Any?>()
+        for (branch in expression.branches) {
+            when (val condition = branch.condition) {
+                is FirEqualityOperatorCall -> {
+                    val arguments = condition.arguments
+                    if (arguments.size == 2 && arguments[0] is FirWhenSubjectExpression) {
+                        val value = when (val targetExpression = arguments[1]) {
+                            is FirConstExpression<*> -> targetExpression.value
+                            is FirQualifiedAccessExpression -> targetExpression.calleeReference.toResolvedCallableSymbol()?.fir as? FirEnumEntry
```

Comment:
`FirEnumEntry` check seems too strict. What if a variable is used multiple times, e.g.,
```kt
val x = ...
vhen (p) {
  x -> ...
  y -> ...
  <!>x<!> -> ...
}
```

Reply:
According to https://punzki.github.io/Kotlin-DiagnosticTestsDiff/2021-05-25/diff-tests__when__When.kt.html, we may need to retrieve the property initializer as well. (Maybe just leaving TODO for now)

```kt
val s = ""
when (x) {
  s -> ...
  <!>""<!> -> ...
}
```

Reply:
Yeah, I guess it depends on where do we draw the line for what an "evaluator" is supposed to do. The above two cases are all doable, but both are non-trivial to define the right scope and implement correctly. For example,
1. for handling general qualified access, we need to know if the property is delegated or not. And we can even do control flow analysis and determine if a mutable property has been modified when evaluating the branch condition, etc. 
2. For the second case, tracing a property access is essentially part of doing constant evaluator. We can only report if the traced result is compile time constant. And we may need to recursively trace the results, etc.

I was under the impression that we won't do all these complex analysis for this diagnostics. Hi, @semoro what's your opinion on this?

Reply:
Discussed with JetBrains and we decided to keep this logic as simple as possible, as it stands now.